File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps):
    shortest_path_len = -1
    row = len(maps)
    col = len(maps[0])
    visited = [[False]*col for _ in range(row)]
    
    delta = [(1,0), (-1,0), (0,1), (0,-1)] 
    
    queue = deque()
    queue.append((0,0,1))
    visited[0][0] = True
    
    while queue :
        cur_r, cur_c, cur_len = queue.popleft()
        if cur_r == row-1 and cur_c == col-1 :
            shortest_path_len = cur_len
            break
        #연결되어 있는 vertex 확인
        for dr, dc in delta:
            next_r = cur_r + dr
            next_c = cur_c + dc
            if next_r >= 0 and next_r < row and next_c >= 0 and next_c < col :
                if maps[next_r][next_c] == 1 and not visited[next_r][next_c] :
                    queue.append((next_r,next_c,cur_len+1))
                    visited[next_r][next_c] = True
    return shortest_path_len
    
def solution(maps):
    return bfs(maps)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;


class Solution {
    static int[][] maps;
    static int answer;
    
    static int[] dr = {-1,1,0,0};
    static int[] dc = {0,0,-1,1};
            
    
    public int solution(int[][] maps) {
        answer = 0;
        this.maps = maps;
        
        int row = maps.length;
        int col = maps[0].length;
        
        if(maps[0][0] != 1 || maps[row-1][col-1] != 1) return -1;
        
        Queue<int[]>queue = new LinkedList<>();
        queue.offer(new int[]{0,0,1});
    
        boolean[][] visited = new boolean[row][col];
        visited[0][0] = true;
        
        while(!queue.isEmpty()) {
            int[] cur = queue.poll();
            int cur_r = cur[0], cur_c = cur[1], cur_len = cur[2];
            
            if (cur_r == row-1 && cur_c == col-1) 
                return cur_len;
            
            for(int i=0;i<4;i++) {
                int next_r = cur_r + dr[i];
                int next_c = cur_c + dc[i];
                
                if(0<=next_r&&next_r<row && 0<=next_c&&next_c<col && !visited[next_r][next_c]
                  && maps[next_r][next_c] == 1) {
                    queue.add(new int[]{next_r, next_c, cur_len+1});
                    visited[next_r][next_c] = true;
                }
            }
        }
        
        
        return -1;
    }
}